### PR TITLE
make mint functions payable

### PR DESCRIPTION
**Pull request by etherinsight**
Other major change: 
- added a withdraw function callable by the owner only to withdraw funds. 

Minor change: 
-  change starting IPFS base URI

### CODE REVIEW

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol [hunk 1]
```diff
@@ -49,6 +50,8 @@ contract PixelAvatars is ERC721Enumerable, ReentrancyGuard, Ownable {
         _;
     }
 
+    uint256 public mintPrice = 0.01 ether;
+
```
Comment (carlomigueldy) [inline pixel-avatar-hardhat/contracts/PixelAvatars.sol:54]: Move this `mintPrice` property next to `_devDaoContract` property and before `validDevDaoToken` modifier. 

By having properties placed right above everything else would make it easier to find smart contract properties

**Review by carlomigueldy — CHANGES_REQUESTED** [review]
Looks good @etherinsight ! 🚀 

I have a few changes requested, and these are just for Solidity common conventions that I know at least.

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol [hunk 2]
```diff
@@ -79,11 +84,31 @@ contract PixelAvatars is ERC721Enumerable, ReentrancyGuard, Ownable {
         emit LogTokenMinted(msg.sender, _devDaoTokenId);
     }
 
+    // provide Mint price in wei
+    function setmintPrice(uint256 newPrice) public onlyOwner {
+        if (newPrice == 0) {
+            mintPrice = newPrice;
+        } else {
+            // TODO: do we need safemath division here?
```
[the remark this answers is not in the document]
Comment (etherinsight) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:92]: I was using division earlier so had left this comment then. There's no more division left so will resolve this.

[the remark this answers is not in the document]
Comment (etherinsight) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:92]: will remove the comment form contract code

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol [hunk 3]
```diff
@@ -79,11 +84,31 @@ contract PixelAvatars is ERC721Enumerable, ReentrancyGuard, Ownable {
         emit LogTokenMinted(msg.sender, _devDaoTokenId);
     }
 
+    // provide Mint price in wei
+    function setmintPrice(uint256 newPrice) public onlyOwner {
```
[the remark this answers is not in the document]
Comment (etherinsight) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:88]: done!

[the remark this answers is not in the document]
Comment (carlomigueldy) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:88]: Awesome!

**Comment by etherinsight** [comment]
Thanks for the review @carlomigueldy. Will update the PR soon. Successfully verified the methods work after change in Etherscan. https://rinkeby.etherscan.io/address/0x8F0EEfABBF5164453D08081c063c1408d08f8c7f#writeContract

**Comment by etherinsight** [comment]
@carlomigueldy How does it look now? Can we ask @Dhaiwat10 to merge it ?

**Review by carlomigueldy — APPROVED** [review]
LGTM! @etherinsight and yup we can merge this in!

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol [hunk 4]
```diff
@@ -79,11 +84,26 @@ contract PixelAvatars is ERC721Enumerable, ReentrancyGuard, Ownable {
...
     function multiMintWithDevDaoToken(uint256[] memory _devDaoTokenIds)
         public
+        payable
         nonReentrant
         multipleDevDaoTokenOwnerOf(_devDaoTokenIds)
     {
+        require(
+            (mintPrice * _devDaoTokenIds.length) <= msg.value,
+            "Ether value sent is not correct"
+        );
```
Comment (Dhaiwat10) [inline pixel-avatar-hardhat/contracts/PixelAvatars.sol:105]: Can't understand this line. Why are we multiplying mintPrice with something and not simply comparing mintPrice with msg.value?

**Comment by carlomigueldy** [comment]
Btw one last request and it is trivial. I forgot to bump the Solidity version back to `^0.8.9` on my previous PR. Can you bump it up to the latest version?

Though I downgraded it to `^0.8.0` because I was not getting better error messages or stack traces as Hardhat gave me that warning and when I was debugging when attempting to write those mocks.

But what are your thoughts on us using the latest Solidity version?

**Review by Dhaiwat10 — CHANGES_REQUESTED** [review]

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol (hunk 4 above)
Comment (carlomigueldy) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:105]: @Dhaiwat10  I think due to the function being able to mint multiple avatars in a single transaction is how we came up with multiplying the `mintPrice` by the given length of `_devDaoTokenIds`

On second thought, perhaps this should match the `msg.value` for the calculated value? @etherinsight 
```
require(
  (mintPrice * _devDaoTokenIds.length) == msg.value, 
  "Ether value sent is not correct"
);
```

**Review by carlomigueldy — CHANGES_REQUESTED** [review]
I just noticed something about the validation/require statements @etherinsight and I think it should match the actual `mintPrice`

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol (hunk 4 above)
Comment (Dhaiwat10) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:105]: I'm sorry but I'm struggling to follow you guys here. Wouldn't `_devDaoTokenIds.length` always equal `8000`?

Comment (Dhaiwat10) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:105]: Oh, never mind. I get it now.

Comment (Dhaiwat10) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:105]: I thought this was the `mint` function and not the `multiMint` function. This makes sense.

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol [hunk 5]
```diff
@@ -61,15 +64,17 @@ contract PixelAvatars is ERC721Enumerable, ReentrancyGuard, Ownable {
...
 
     /// @dev Private sale minting (reserved for DevDao owners)
     function mintWithDevDaoToken(uint256 _devDaoTokenId)
         public
+        payable
         nonReentrant
         validDevDaoToken(_devDaoTokenId)
         devDaoTokenOwnerOf(_devDaoTokenId)
     {
+        require(mintPrice <= msg.value, "Not enough ether sent");
```
[the remark this answers is not in the document]
Comment (Dhaiwat10) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:77]: I guess having `<=` is fine. We don't mind if someone sends extra ether in the form of donations haha

**Review by Dhaiwat10 — APPROVED** [review]

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol (hunk 4 above)
Comment (carlomigueldy) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:105]: I see no worries @Dhaiwat10 . `_devDaoTokenIds` references to the function argument for `multiMintWithDevDaoToken`

So it's coming from a User, say is User owns multiple DevDAO tokens `6300` and `2489` then he would input `"[6300, 2489]"` on an input text field like the one from EtherScan > Contract > Write

Comment (carlomigueldy) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:105]: Oh. Yes @Dhaiwat10

File: pixel-avatar-hardhat/contracts/PixelAvatars.sol (hunk 5 above)
[the remark this answers is not in the document]
Comment (carlomigueldy) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:77]: Yeah 🤔 who wouldn't love donations XD

[the remark this answers is not in the document]
Comment (etherinsight) [reply pixel-avatar-hardhat/contracts/PixelAvatars.sol:77]: Yeah that's the idea for now. Let the devs send extra donation if they want to

**Pull request merged**